### PR TITLE
feat: PortDirection field with inferDirection defaults

### LINKED ISSUE #1930: feat: add PortDirection field to InterfaceTemplate and PlacedPort
## Summary

Add a `direction` field (`input` | `output` | `bidirectional`) to InterfaceTemplate and PlacedPort to support AV signal flow direction.

## Acceptance Criteria

- [ ] Add `PortDirectionSchema` to schemas (`input`, `output`, `bidirectional`)
- [ ] Add `direction?: PortDirection` to InterfaceTemplate (optional, defaults inferred)
- [ ] Add `direction?: PortDirection` to PlacedPort (optional, overrides template default)
- [ ] Add `inferDirection(type, mgmtOnly)` utility in `port-utils.ts`:
  - `mgmt_only: true` → `input`
  - `console`, `rs-232`, `rs-422` types → `input`
  - All AV types (XLR, HDMI, SDI, etc.) → no default (explicit direction required)
  - Everything else → `bidirectional`
- [ ] PortIndicators shows direction (arrow indicators for input/output, none for bidirectional)
- [ ] PortTooltip shows direction label
- [ ] Existing layouts without direction field load correctly (backward compatible)
- [ ] Schema passthrough preserves unknown fields (already in place)

## Technical Notes

- Direction defaults to `bidirectional` for network ports (backward compatible).
- AV ports should require explicit direction in device type definitions.
- The `inferDirection()` utility provides smart defaults so existing device types without direction still work.
- Follow the same pattern as `label`: InterfaceTemplate provides default, PlacedPort can override.
- PortIndicators renders from templates. Its grouped mode (over 24 ports) shows count badges, so per-port direction arrows will not be visible in that mode.
- An upgrade-corpus fixture with a direction-bearing port is welcome but not mandatory, since `direction` is an additive optional field.

## Related

- Epic #1928
- Spike #1927 findings
- Depends on #1929: the `inferDirection` defaults reference types (`rs-232`, `rs-422`, and the AV types) that only exist after #1929 lands.

### PR DESCRIPTION
**Pull request by ggfevans**
## Summary

Adds an optional `direction` field (`input` | `output` | `bidirectional`) to `InterfaceTemplate` and `PlacedPort`, so AV signal flow direction can be modeled and shown in the UI.

## Field semantics

`direction` follows the exact same pattern as the existing `label` field: `InterfaceTemplate.direction` is the device-type default, and `PlacedPort.direction` can independently override it for a specific placed instance. Both fields are optional and additive; existing layouts without `direction` continue to load unchanged (schema passthrough already preserves unknown fields, and this is a new optional field, not a required one).

## inferDirection(type, mgmtOnly)

New utility in `src/lib/utils/port-utils.ts`, used wherever a template or placed port omits an explicit `direction`:

| Input | Result |
| --- | --- |
| `mgmt_only: true` | `input` |
| `type` is `console`, `rs-232`, or `rs-422` | `input` |
| `type` is any other AV type in `AV_INTERFACE_TYPES` | `undefined` (no default; direction must be set explicitly) |
| everything else (network, power, USB, etc.) | `bidirectional` |

Check order matters: `rs-232` and `rs-422` are members of both the input-default list and `AV_INTERFACE_TYPES` (added by #1929's control-type taxonomy). The input-default check runs first, so those two resolve to `input` rather than falling through to "no default".

## UI

- `PortIndicators.svelte`: individual (low-density, <=24 ports) port circles now show a small left/right arrow for `input`/`output`; `bidirectional` and unset (AV types without an explicit direction) show nothing. The resolved direction is `iface.direction ?? inferDirection(iface.type, iface.mgmt_only)`. Grouped high-density badge mode (>24 ports) is unchanged, per the issue: per-port arrows aren't visible there, and that's accepted. This change is deliberately scoped to arrow rendering only, since #3089 will refactor `PortIndicators.svelte` after this merges.
- `PortTooltip.svelte`: shows a human-readable direction label ("Input" / "Output" / "Bidirectional") under the port type, using the same resolution. No label is shown when direction is neither set nor inferable.

## Schema

- `PortDirectionSchema` added to `src/lib/schemas/index.ts`, dual-sourced with a hand-written `PortDirection` type in `src/lib/types/index.ts` (existing pattern for this file pair).
- `static/schemas/rackula-layout.schema.json` regenerated via `npm run generate-schema` (byte-for-byte, verified by `layout-json-schema.test.ts`).
- No `SCHEMA_VERSION` bump: it's already `1.1` this release (#3087) and `direction` is an additive optional field on both schemas.

## Corpus fixture

Added `src/tests/fixtures/upgrade-corpus/v26.7.0-port-direction.rackula.yaml`: an interface template with an explicit `direction: output`, a placed port for that same interface overriding it to `direction: input`, and a plain network interface with no `direction` at all (covering the backward-compatible/optional path). `schema_version: "1.0"` in the fixture metadata, matching sibling fixtures.

The sidecar allow-lists `\.position$`: the fixture's device `position: 3` gets converted to internal 1/6U units on load, which the corpus checker would otherwise flag as data loss. The sibling `v26.7.0-av-connectivity-ports` fixture avoids this only by coincidence (its device's raw position value happens to also appear as a `template_index` elsewhere in the layout); this fixture uses the explicit, documented allow-list mechanism instead (same pattern as `over-rack-rail-position.expected.json`).

## Test evidence

- `VITEST_MAX_WORKERS=2 npm run test:run`: 248 test files, 3653 tests, all passing.
- `npm run check` (svelte-check): 0 errors, 0 warnings across 5504 files.
- `npx eslint` on all touched files: no issues.
- `src/tests/port-utils.test.ts` (new): covers every `inferDirection` branch (mgmt_only, console/rs-232/rs-422, all AV types, network/power/USB/management/virtual defaults, and the mgmt_only-over-AV precedence edge case), plus a schema-level test that a `PlacedPort.direction` override can differ from its `InterfaceTemplate.direction` default.
- `src/tests/upgrade-corpus.test.ts`: new fixture passes with no undeclared silent data loss.
- `bash scripts/check-corpus-freshness.sh v26.7.0`: passes.
- `svelte-autofixer` (Svelte MCP) on both modified `.svelte` files: no issues from this change (PortIndicators surfaces one pre-existing, unrelated suggestion about its `$effect`-based hover timeout cleanup, not touched here).

## Deviations from the issue

No PlacedPort/InterfaceTemplate merge helper exists in the codebase for `label` either (`PortIndicators`/`PortTooltip` currently render only from `InterfaceTemplate`, and `instantiatePorts()` doesn't copy `label` onto the placed port). Since there's no existing merge utility to mirror, this PR doesn't invent one for `direction`; it adds the field with the same override-capable shape as `label` and leaves resolution to whatever future consumer wires `PlacedPort` into the UI (tracked separately, out of scope here).

Closes #1930

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01AadzUkj8Q4YcAjXGeoG9b2

___

## **CodeAnt-AI Description**
**Show port signal direction in the layout and tooltip**

### What Changed
- Ports can now carry a direction of input, output, or bidirectional on both device templates and placed ports
- When a direction is not set, the app now infers a default for common cases like management-only and console ports, while leaving some AV ports without a default
- Small port arrows now appear for input and output ports in the device view, and the tooltip shows a plain-language direction label
- Added support data and tests for direction-aware ports, including a sample layout fixture

### Impact
`✅ Clearer AV signal flow`
`✅ Fewer ambiguous port labels`
`✅ Easier port review in tooltips`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: src/lib/utils/port-utils.ts [hunk 1]
```diff
@@ -42,6 +42,14 @@ const AV_INTERFACE_TYPES = new Set<string>([
   "dante",
 ]);
 
+/**
+ * Interface types whose direction defaults to "input" (spike #1927 taxonomy):
+ * management/console ports and the AV control-serial types. Checked before
+ * AV_INTERFACE_TYPES in inferDirection() since rs-232/rs-422 are members of
+ * both sets and the specific input default takes priority.
+ */
+const INPUT_DEFAULT_TYPES = new Set<string>(["console", "rs-232", "rs-422"]);
```
[the remark this answers is not in the document]
Comment (ggfevans) [reply src/lib/utils/port-utils.ts:54]: Confirmed: the code was already correct and tested (inferDirection("management") -> "bidirectional" via the test at src/tests/port-utils.test.ts, and inferDirection("1000base-t", true) -> "input" via the mgmt_only branch). Only the comment was misleading, since it read as if the "management" interface type were a member of INPUT_DEFAULT_TYPES when it isn't; management-only routing goes through the separate mgmt_only check. Reworded the comment to make that explicit, no behavior change. Fixed in 5beb9d3f.

File: src/lib/components/PortIndicators.svelte [hunk 2]
```diff
@@ -214,6 +227,27 @@
...
+        <!-- Direction arrow (input/output only; bidirectional and unset show nothing) -->
+        {#if getPortDirection(iface) === "input"}
+          <text
+            class="port-direction-indicator"
+            x={x - PORT_RADIUS - 2}
+            y={y + 2}
+            text-anchor="end"
+          >
+            &#8592;
+          </text>
```
[the remark this answers is not in the document]
Comment (ggfevans) [reply src/lib/components/PortIndicators.svelte:240]: Reproduced this against the actual SVG rendering: the input arrow was &#8592; (left-pointing) positioned at x - PORT_RADIUS - 2 with text-anchor="end", so with text-anchor="end" placing the glyph's trailing edge at that x, the arrowhead pointed further left, away from the port, on the port's left side. Visually that reads the same as "outward flow from the left side", which is the wrong signal for input. Applied the suggested fix: swapped the input glyph to &#8594; (right-pointing) so, with the same position/text-anchor, the arrowhead lands at the port edge and points into it, matching the existing output arrow's outward-pointing convention on the other side. Fixed in 5beb9d3f.

File: src/lib/schemas/index.ts [hunk 3]
```diff
@@ -1191,6 +1199,7 @@ export type InterfaceType = z.infer<typeof InterfaceTypeSchema>;
 export type PoEType = z.infer<typeof PoETypeSchema>;
 export type PoEMode = z.infer<typeof PoEModeSchema>;
 export type InterfacePosition = z.infer<typeof InterfacePositionSchema>;
+export type PortDirection = z.infer<typeof PortDirectionSchema>;
```
[the remark this answers is not in the document]
Comment (ggfevans) [reply src/lib/schemas/index.ts:1202]: Evaluated on merit and declining this one. The z.infer re-export in schemas/index.ts isn't specific to this PR's PortDirection addition; it's the file's established pattern for every enum schema (InterfacePosition, PoEMode, DeviceCategory, FormFactor, InterfaceType, PoEType, and about a dozen more all get the same `export type X = z.infer<typeof XSchema>` line immediately after their hand-written counterpart in src/lib/types/index.ts). None of those are currently imported from $lib/schemas by name either (verified by grep; only the *Schema objects are imported from there, e.g. DeviceCategorySchema, FormFactorSchema), so PortDirection is consistent with, not a departure from, the surrounding code. Special-casing PortDirection alone would make it the one enum in this block without a schemas-side type export, which is a worse inconsistency than the redundancy being flagged. If the project wants to prune this dual-export pattern, that's a repo-wide cleanup across ~15+ existing lines, out of scope for this PR. Leaving as-is.
